### PR TITLE
Initialize resource via DI

### PR DESCRIPTION
**Pull request by Ardagan**
**What changed?**
Initialize resource via DI.

**Why?**
Migrate to Uber/Fx

**How did you test it?**
local/automated tests

**Potential risks**
History service doesn't initialize correctly.

**Is hotfix candidate?**
No

### CODE REVIEW

File: temporal/server.go [hunk 1]
```diff
@@ -194,18 +198,32 @@ func (s *Server) Start() error {
...
+				),
+				history.Module)
 			err = histApp.Err()
-			svc = historySvc
+			if err != nil {
+				close(s.serviceStoppedChs[svcName])
+				return fmt.Errorf("unable to construct service %q: %w", svcName, err)
+			}
+			s.serviceApps[svcName] = histApp
+			err = histApp.Start(ctx.Background())
```
Comment (yiminc) [inline temporal/server.go:221]: Do you think we need some reasonable timeout here to prevent this call from hang?

File: temporal/server.go [hunk 2]
```diff
@@ -194,18 +198,32 @@ func (s *Server) Start() error {
...
 		case primitives.HistoryService:
-			var historySvc *history.Service
+			// todo: generalize this custom case logic as other services onboard fx
 			histApp := fx.New(
-				fx.Supply(params),
-				history.Module,
-				fx.Populate(&historySvc))
+				fx.Supply(
+					params,
+					s.serviceStoppedChs[svcName],
```
Comment (yiminc) [inline temporal/server.go:212]: this simply supply a channel, how is this wired up? where does it close this channel when service close?

Comment (Ardagan) [reply temporal/server.go:212]: Used by lifetime hooks. Currently it's generic, but I want to keep it simple for now and solve collisions further down the line.

File: common/resource/fx.go [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+	fx.Provide(ClusterMetadataManagerProvider),
+	fx.Provide(PersistenceServiceResolverProvider),
+	fx.Provide(PersistenceMaxQpsProvider),
+	fx.Provide(AbstractDatastoreFactoryProvider),
+	fx.Provide(ClusterNameProvider),
+	fx.Provide(MetricsClientProvider),
+	persistenceClient.FactoryModule,
+	fx.Provide(SAProviderProvider),
+	fx.Provide(SAManagerProvider),
+	fx.Provide(SAMapperProvider),
```
Comment (yiminc) [inline common/resource/fx.go:89]: prefer to expand SA -> SearchAttribute

File: host/onebox.go [hunk 4]
```diff
@@ -477,9 +477,11 @@ func (c *temporalImpl) startHistory(
 			}
 		}
 
+		stoppedCh := make(chan struct{})
 		var historyService *history.Service
 		app := fx.New(
-			fx.Supply(params),
+			fx.Supply(params,
```
Comment (yiminc) [inline host/onebox.go:483]: nit: put params in new line

File: temporal/server.go [hunk 5]
```diff
@@ -194,18 +199,34 @@ func (s *Server) Start() error {
...
+				),
+				history.Module)
 			err = histApp.Err()
-			svc = historySvc
+			if err != nil {
+				close(s.serviceStoppedChs[svcName])
+				return fmt.Errorf("unable to construct service %q: %w", svcName, err)
+			}
+			s.serviceApps[svcName] = histApp
+			timeoutCtx, cancelFunc := context.WithTimeout(ctx.Background(), time.Duration(15)*time.Second)
```
Comment (yiminc) [inline temporal/server.go:222]: define const for startupTimeout which will be used by other service as well.

**Review by yiminc — APPROVED** [review]

**Pull request merged**